nvflare/private/fed/server/fed_server.py

```diff
@@ -596,28 +599,60 @@ def AuxCommunicate(self, request, context):
+            command_shareable = Shareable()
+            command_shareable.set_header(ServerCommandKey.PEER_FL_CONTEXT, shared_fl_context)
+            command_shareable.set_header(ServerCommandKey.TOPIC, topic)
+            command_shareable.set_header(ServerCommandKey.SHAREABLE, shareable)
+            data = {
+                ServerCommandKey.COMMAND: ServerCommandNames.AUX_COMMUNICATE,
+                ServerCommandKey.DATA: command_shareable,
+            }
+            if shareable.get_header(CollectiveCommShareableHeader.IS_COLLECTIVE_AUX, False):
+                with self.engine.lock:
+                    job_id = str(shared_fl_context.get_prop(FLContextKey.CURRENT_RUN))
+                    command_conn = self.engine.get_collective_command_conn(job_id)
+                    participate_clients = len(self.engine.run_processes[job_id][RunProcessKey.PARTICIPANTS])
+                    command_waiter = self.engine.run_processes[job_id][RunProcessKey.COLLECTIVE_CONNECTION_WAITER]
+                    self.logger.info(f"participate client num is {participate_clients}")
                 if command_conn:
-                    command_shareable = Shareable()
-                    command_shareable.set_header(ServerCommandKey.PEER_FL_CONTEXT, shared_fl_context)
-                    command_shareable.set_header(ServerCommandKey.TOPIC, topic)
-                    command_shareable.set_header(ServerCommandKey.SHAREABLE, shareable)
-
-                    data = {
-                        ServerCommandKey.COMMAND: ServerCommandNames.AUX_COMMUNICATE,
-                        ServerCommandKey.DATA: command_shareable,
-                    }
-                    command_conn.send(data)
-
-                    return_data = command_conn.recv()
-                    reply = return_data.get(ServerCommandKey.AUX_REPLY)
-                    child_fl_ctx = return_data.get(ServerCommandKey.FL_CONTEXT)
-
-                    fl_ctx.props.update(child_fl_ctx)
+                    shareable.set_header(CollectiveCommShareableHeader.WORLD_SIZE, participate_clients)
+                    with command_waiter.receive_all:
+                        command_waiter.received_requests += 1
+                        command_conn.send(data)
+                        while True:
+                            if command_waiter.received_requests == participate_clients:
+                                break
+                            if not command_waiter.receive_all.wait(self._collective_comm_timeout):
+                                # Known issue:: IF the number of threads is lower than number of clients (world_size)
+                                #       then this will be DEAD-LOCK
+                                # Assumption: all the client request need to be here to proceed
+                                data[CollectiveCommandKey.TIMEOUT] = f"Timeout for {self._collective_comm_timeout}"
+                                command_conn.send(data)
```

Comment:
if it's a timeout, not receiving the client AuxCommunicate update, should this send to the child process?

Reply:
yes, the child process need to know things are bad so can respond to it